src/main/java/com/aloc/aloc/problem/service/ProblemSolvingService.java

```diff
@@ -61,25 +61,28 @@ public String checkAndUpdateProblemSolved(ProblemResponseDto problem, User user)
 		}
 	}
 
-	private int calculateCoinToAdd(Long problemId) {
-		// 3등까지는 50코인, 4등부터는 30코인을 지급합니다.
-		long solvedUserCount = userProblemRepository.countSolvingUsersByProblemId(problemId, currentSeason);
-		return solvedUserCount <= MAX_SOLVED_USER_COUNT_FOR_BONUS ? COINS_FOR_TOP_SOLVERS : COINS_FOR_OTHER_SOLVERS;
+	public List<UserProblem> getSolvedUserListByProblemId(Long problemId) {
+		return userProblemRepository.findAllByProblemIdAndIsSolvedIsTrue(problemId);
 	}
 
-	private void updateUserAndSaveSolvedProblem(User user, Long problemId) {
+	private void updateUserAndSaveSolvedProblem(User user, Problem problem) {
 		// 코인을 지급하고 사용자 정보를 저장합니다.
-		int coinToAdd = calculateCoinToAdd(problemId);
+		int coinToAdd;
+		if (problem.getProblemType().getRoutine().equals(Routine.DAILY)) {
```

Comment:
```suggestion
	 public void updateUserAndSaveSolvedProblem(User user, Problem problem) {
        int coinToAdd = calculateCoinToAdd(problem, user.getCourse());
        updateUserWithCoin(user, coinToAdd);
        userService.checkUserRank(user);
    }

    private int calculateCoinToAdd(Problem problem, Course course) {
        return problem.getProblemType().getRoutine() == Routine.DAILY
                ? coinService.calculateCoinToAddForDaily(problem.getId())
                : coinService.calculateCoinToAddForWeekly(problem.getAlgorithm(), course);
    }
```

이런 식으로 calculateCoinToAdd와 updateUserAndSaveSolvedProblem 메소드의 책임을 분리하는 것도 좋을 것 같아요. :)

Reply:
오 좋은 것 같아요! 수정하겠습니다!